benchmark/single-source/StringBreadcrumbs.swift

```diff
@@ -0,0 +1,494 @@
+    13 swift                    0x00000001000b38f4 swift::performFrontend(llvm::ArrayRef<char const*>, char const*, void*, swift::FrontendObserver*) + 3028
+    14 swift                    0x000000010006ca44 main + 660
+    15 libdyld.dylib            0x00007fff63e293f1 start + 1
+    16 libdyld.dylib            0x0000000000000008 start + 2619173912
+    Illegal instruction: 4
+
+  """#
+let shortASCIIWorkload = Workload(
+  name: "shortASCII",
+  string: asciiBase)
+let longASCIIWorkload = Workload(
+  name: "longASCII",
+  string: String(repeating: asciiBase, count: 100))
+
+let mixedBase = """
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    👍👩‍👩‍👧‍👧👨‍👨‍👦‍👦🇺🇸🇨🇦🇲🇽👍🏻👍🏼👍🏽👍🏾👍🏿
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    👍👩‍👩‍👧‍👧👨‍👨‍👦‍👦🇺🇸🇨🇦🇲🇽👍🏻👍🏼👍🏽👍🏾👍🏿the quick brown fox👍🏿👍🏾👍🏽👍🏼👍🏻🇲🇽🇨🇦🇺🇸👨‍👨‍👦‍👦👩‍👩‍👧‍👧👍
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    今回のアップデートでSwiftに大幅な改良が施され、安定していてしかも直感的に使うことができるAppleプラットフォーム向けプログラミング言語になりました。
+    Worst thing about working on String is that it breaks *everything*. Asserts, debuggers, and *especially* printf-style debugging 😭
+    Swift 是面向 Apple 平台的编程语言，功能强大且直观易用，而本次更新对其进行了全面优化。
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    이번 업데이트에서는 강력하면서도 직관적인 Apple 플랫폼용 프로그래밍 언어인 Swift를 완벽히 개선하였습니다.
+    Worst thing about working on String is that it breaks *everything*. Asserts, debuggers, and *especially* printf-style debugging 😭
+    в чащах юга жил-был цитрус? да, но фальшивый экземпляр
+    siebenhundertsiebenundsiebzigtausendsiebenhundertsiebenundsiebzig
+    \u{201c}Hello\u{2010}world\u{2026}\u{201d}
+    \u{300c}\u{300e}今日は\u{3001}世界\u{3002}\u{300f}\u{300d}
+    Worst thing about working on String is that it breaks *everything*. Asserts, debuggers, and *especially* printf-style debugging 😭
+
+  """
+let shortMixedWorkload = Workload(
+  name: "shortMixed",
+  string: mixedBase)
+let longMixedWorkload = Workload(
+  name: "longMixed",
+  string: String(repeating: mixedBase, count: 100))
+
+
+//==============================================================================
+// Benchmarks
+//==============================================================================
+
+/// Convert `count` random UTF-16 offsets into String indices.
+class UTF16ToIndex: BenchmarkBase {
+  let count: Int
+  var inputOffsets: [Int] = []
+  var outputIndices: [String.Index] = []
+
+  init(workload: Workload, count: Int) {
+    self.count = count
+    super.init(name: "StringBreadcrumbs.UTF16ToIndex", workload: workload)
+  }
+
+  override func setUp() {
+    super.setUp()
+    var rng = LCRNG(seed: seed)
+    let range = 0 ..< inputString.utf16.count
+    inputOffsets = Array(range.shuffled(using: &rng).prefix(count))
+    outputIndices = []
+    outputIndices.reserveCapacity(inputOffsets.count)
+  }
+
+  override func tearDown() {
+    super.tearDown()
+    inputOffsets = []
+  }
+
+  @inline(never)
+  override func run() {
```

Comment:
Maybe make these `final` to have static dispatch?

Reply:
I don't think that's necessary -- these are non-@objc internal classes so the optimizer should be able to see that there aren't any further overrides. (And the benchmark payload will hopefully be overwhelmingly larger than a virtual method call.)

(Lots of thanks for the run_smoke_bench fix, btw! I wasn't aware we were matching benchmark names like that.)

Reply:
Sure, assuming you've verified this really has no effect...